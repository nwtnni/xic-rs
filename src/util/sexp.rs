use std::borrow::Cow;

use pretty::{BoxDoc, Doc};

use crate::error;
use crate::util::symbol;
use crate::util::Tap;

#[derive(Clone, Debug)]
pub enum Sexp {
    Atom(Cow<'static, str>),
    List(Vec<Sexp>), 
}

impl Sexp {
    fn to_doc(&self) -> Doc<BoxDoc<()>> {
        match self {
        | Sexp::Atom(s) => Doc::as_string(s),
        | Sexp::List(l) => {
            Doc::text("(")
                .append(
                    Doc::intersperse(
                        l.iter().map(Sexp::to_doc),
                        Doc::space(),
                    )
                    .nest(1)
                    .group()
                )
                .append(Doc::text(")"))
        }
        }
    }

    pub fn write<W: std::io::Write>(&self, width: usize, writer: &mut W) -> Result<(), error::Error> {
        self.to_doc().render(width, writer)?;
        Ok(())
    }
}

pub trait Serialize: Sized {
    fn sexp(&self) -> Sexp;
    fn sexp_move(self) -> Sexp { self.sexp() }
}

impl Serialize for Sexp {
    fn sexp(&self) -> Sexp { self.clone() }
    fn sexp_move(self) -> Sexp { self }
}

impl Serialize for i64 {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(self.to_string())) }
}

impl Serialize for symbol::Symbol {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(symbol::resolve(*self))) }
}

impl Serialize for &'static str {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(*self)) }
}

impl Serialize for String {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(self.clone())) }
    fn sexp_move(self) -> Sexp { Sexp::Atom(Cow::from(self)) }
}

impl<T: Serialize> Serialize for Vec<T> {
    fn sexp(&self) -> Sexp {
        self.iter()
            .map(Serialize::sexp)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }

    fn sexp_move(self) -> Sexp {
        self.into_iter()
            .map(Serialize::sexp_move)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }
}
