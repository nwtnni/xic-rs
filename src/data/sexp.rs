use std::borrow::Cow;
use std::fmt;

use pretty::Arena;
use pretty::DocAllocator;
use pretty::DocBuilder;

use crate::error;
use crate::data::symbol;
use crate::util::Tap;

#[derive(Clone, Debug)]
pub enum Sexp {
    Atom(Cow<'static, str>),
    List(Vec<Sexp>),
}

impl fmt::Display for Sexp {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        let arena = Arena::new();
        self.to_doc(&arena).render_fmt(80, fmt)
    }
}

impl Sexp {
    fn to_doc<'a, A>(&self, allocator: &'a A) -> DocBuilder<'a, A, ()>
    where
        A: DocAllocator<'a, ()>,
        A::Doc: Clone,
    {
        match self {
            Sexp::Atom(atom) => allocator.text(atom.clone()),
            Sexp::List(list) => allocator
                .intersperse(
                    list.iter().map(|sexp| sexp.to_doc(allocator).nest(4)),
                    allocator.line(),
                )
                .parens()
                .group(),
        }
    }

    pub fn write<W: std::io::Write>(
        &self,
        width: usize,
        writer: &mut W,
    ) -> Result<(), error::Error> {
        let arena = Arena::new();
        self.to_doc(&arena).render(width, writer)?;
        Ok(())
    }
}

pub trait Serialize: Sized {
    fn sexp(&self) -> Sexp;

    fn sexp_move(self) -> Sexp {
        self.sexp()
    }
}

impl Serialize for Sexp {
    fn sexp(&self) -> Sexp {
        self.clone()
    }

    fn sexp_move(self) -> Sexp {
        self
    }
}

impl Serialize for i64 {
    fn sexp(&self) -> Sexp {
        Sexp::Atom(Cow::from(self.to_string()))
    }
}

impl Serialize for symbol::Symbol {
    fn sexp(&self) -> Sexp {
        Sexp::Atom(Cow::from(symbol::resolve(*self)))
    }
}

impl Serialize for &'static str {
    fn sexp(&self) -> Sexp {
        Sexp::Atom(Cow::from(*self))
    }
}

impl Serialize for String {
    fn sexp(&self) -> Sexp {
        Sexp::Atom(Cow::from(self.clone()))
    }

    fn sexp_move(self) -> Sexp {
        Sexp::Atom(Cow::from(self))
    }
}

impl<T: Serialize> Serialize for Vec<T> {
    fn sexp(&self) -> Sexp {
        self.iter()
            .map(Serialize::sexp)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }

    fn sexp_move(self) -> Sexp {
        self.into_iter()
            .map(Serialize::sexp_move)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }
}

impl<const N: usize, T: Serialize> Serialize for [T; N] {
    fn sexp(&self) -> Sexp {
        self.iter()
            .map(Serialize::sexp)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }

    fn sexp_move(self) -> Sexp {
        IntoIterator::into_iter(self)
            .map(Serialize::sexp_move)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }
}
