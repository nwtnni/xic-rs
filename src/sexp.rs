use std::borrow::Cow;

use crate::symbol;
use crate::util::Tap;

#[derive(Clone, Debug)]
pub enum Sexp {
    Atom(Cow<'static, str>),
    List(Vec<Sexp>), 
}

pub trait Serialize: Sized {
    fn sexp(&self) -> Sexp;
    fn sexp_move(self) -> Sexp { self.sexp() }
}

impl Serialize for Sexp {
    fn sexp(&self) -> Sexp { self.clone() }
    fn sexp_move(self) -> Sexp { self }
}

impl Serialize for symbol::Symbol {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(symbol::resolve(*self))) }
}

impl Serialize for &'static str {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(*self)) }
}

impl Serialize for String {
    fn sexp(&self) -> Sexp { Sexp::Atom(Cow::from(self.clone())) }
    fn sexp_move(self) -> Sexp { Sexp::Atom(Cow::from(self)) }
}

impl<T: Serialize> Serialize for Vec<T> {
    fn sexp(&self) -> Sexp {
        self.iter()
            .map(Serialize::sexp)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }

    fn sexp_move(self) -> Sexp {
        self.into_iter()
            .map(Serialize::sexp_move)
            .collect::<Vec<_>>()
            .tap(Sexp::List)
    }
}
