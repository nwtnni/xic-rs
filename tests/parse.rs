use std::fmt;
use std::path::Path;

use xic::util::sexp::Serialize as _;

struct Snapshot(Result<xic::data::ast::Program, xic::Error>);

impl fmt::Display for Snapshot {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        match &self.0 {
            Ok(program) => {
                let mut buffer = Vec::new();
                program.sexp().write(80, &mut buffer).unwrap();
                write!(fmt, "{}", std::str::from_utf8(&buffer).unwrap())
            }
            Err(error) => write!(fmt, "{}", error),
        }
    }
}

#[test_generator::test_resources("tests/parse/*.xi")]
pub fn parse(path: &str) {
    let lexer = xic::lex::Driver::new(Path::new("."), false);
    let parser = xic::parse::Driver::new(Path::new("."), false);

    let lexed = lexer.drive(Path::new(path)).unwrap();
    let parsed = parser.drive(Path::new(path), lexed);

    insta::assert_display_snapshot!(path, Snapshot(parsed));
}
