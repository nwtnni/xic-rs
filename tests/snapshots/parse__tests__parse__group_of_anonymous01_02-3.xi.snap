---
source: tests/parse.rs
assertion_line: 29
expression: Snapshot(parsed)
---
(() ((main () () ((f) (if a ((return))) ((return)) ((f)) ((())) (return)))))
